File: Sources/Minutes/Providers/OAuthService.swift
import Foundation
import MinutesCore

/// OAuth PKCE sign-in, token storage and refresh for Codex and Grok.
actor OAuthService {
    private static let redirectTimeout: Duration = .seconds(300)

    private let keychain = TokenKeychain()
    private var pending: [ProviderID: PKCE] = [:]
    private var cache: [ProviderID: TokenBundle] = [:]

    func isConnected(_ provider: ProviderID) -> Bool {
        (try? storedTokens(provider)) != nil
    }

    /// Starts a sign-in and returns the URL to open in the browser.
    func begin(_ provider: ProviderID) -> URL {
        let pkce = PKCE.generate()
        pending[provider] = pkce
        return OAuthConfig.config(for: provider).authorizeURL(pkce: pkce)
    }

    /// Waits for the browser redirect on the provider's loopback port.
    nonisolated func awaitRedirect(_ provider: ProviderID) async throws -> OAuthCallback {
        let config = OAuthConfig.config(for: provider)
        let server = try LoopbackServer(port: config.callbackPort, path: config.callbackPath)
        return try await withThrowingTaskGroup(of: OAuthCallback.self) { group in
            group.addTask { try await server.waitForCallback() }
            group.addTask {
                try await Task.sleep(for: Self.redirectTimeout)
                throw OAuthError.providerError("Timed out waiting for the browser.")
            }
            defer { group.cancelAll() }
            return try await group.next()!
        }
    }

    /// Exchanges the authorization code for tokens and stores them.
    func complete(_ provider: ProviderID, callback: OAuthCallback) async throws {
        guard let pkce = pending[provider] else { throw OAuthError.providerError("Start connecting first.") }
        if let state = callback.state, state != pkce.state { throw OAuthError.stateMismatch }
        let config = OAuthConfig.config(for: provider)
        let data = try await postForm(config.tokenEndpoint, config.exchangeParameters(code: callback.code, verifier: pkce.verifier))
        let bundle = try TokenBundle.from(tokenResponse: data, previous: nil, now: Date(), expirySkew: config.expirySkew)
        try store(bundle, for: provider)
        pending[provider] = nil
    }

    func disconnect(_ provider: ProviderID) throws {
        cache[provider] = nil
        pending[provider] = nil
        try keychain.delete(provider)
    }

    /// A valid token bundle, refreshed if it is about to expire.
    func activeTokens(_ provider: ProviderID) async throws -> TokenBundle {
        guard let tokens = try storedTokens(provider) else { throw SummaryError.noProviderConnected }
        guard tokens.needsRefresh(now: Date()) else { return tokens }
        let config = OAuthConfig.config(for: provider)
        do {
            let data = try await postForm(config.tokenEndpoint, config.refreshParameters(refreshToken: tokens.refreshToken))
            let refreshed = try TokenBundle.from(tokenResponse: data, previous: tokens, now: Date(), expirySkew: config.expirySkew)
            try store(refreshed, for: provider)
            return refreshed
        } catch let error as OAuthError {
            // The refresh token was rejected: the user has to sign in again.
            try? disconnect(provider)
            throw SummaryError.provider("\(provider.displayName) needs to be reconnected in Settings. \(error.localizedDescription)")
        }
    }

    private func storedTokens(_ provider: ProviderID) throws -> TokenBundle? {
        if let cached = cache[provider] { return cached }
        let loaded = try keychain.load(provider)
        cache[provider] = loaded
        return loaded
    }

    private func store(_ bundle: TokenBundle, for provider: ProviderID) throws {
        try keychain.save(bundle, for: provider)
        cache[provider] = bundle
    }

    private func postForm(_ url: URL, _ parameters: [String: String]) async throws -> Data {
        var components = URLComponents()
        components.queryItems = parameters.map { URLQueryItem(name: $0.key, value: $0.value) }
        var request = URLRequest(url: url, timeoutInterval: 30)
        request.httpMethod = "POST"
        request.setValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        // URLComponents leaves "+" unescaped, which form decoding would read as a space.
        request.httpBody = Data((components.percentEncodedQuery ?? "").replacingOccurrences(of: "+", with: "%2B").utf8)
        let (data, _) = try await URLSession.shared.data(for: request)
        return data
    }
}

File: src-tauri/src/providers/oauth_service.rs
//! OAuth PKCE sign-in, token storage and refresh for Codex and Grok.
use std::collections::HashMap;
use std::sync::Mutex;
use std::time::Duration;

use chrono::Utc;

use super::{NetResult, keychain};
use super::loopback;
use crate::core::oauth::{OAuthCallback, OAuthConfig, Pkce, ProviderId, TokenBundle};
use crate::core::{Error, Result};

const REDIRECT_TIMEOUT: Duration = Duration::from_secs(300);
const REQUEST_TIMEOUT: Duration = Duration::from_secs(30);

pub struct OAuthService {
    http: reqwest::Client,
    pending: Mutex<HashMap<ProviderId, Pkce>>,
    cache: Mutex<HashMap<ProviderId, TokenBundle>>,
    /// One refresh at a time, so parallel requests do not spend the same refresh token twice.
    refreshing: tokio::sync::Mutex<()>,
}

impl OAuthService {
    pub fn new(http: reqwest::Client) -> Self {
        Self { http, pending: Mutex::default(), cache: Mutex::default(), refreshing: tokio::sync::Mutex::new(()) }
    }

    pub fn is_connected(&self, provider: ProviderId) -> bool {
        matches!(self.stored_tokens(provider), Ok(Some(_)))
    }

    /// Starts a sign-in and returns the URL to open in the browser.
    pub fn begin(&self, provider: ProviderId) -> String {
        let pkce = Pkce::generate();
        let url = OAuthConfig::for_provider(provider).authorize_url(&pkce);
        self.pending.lock().unwrap().insert(provider, pkce);
        url
    }

    /// Waits for the browser redirect on the provider's loopback port.
    pub async fn await_redirect(provider: ProviderId) -> Result<OAuthCallback> {
        let config = OAuthConfig::for_provider(provider);
        tokio::time::timeout(REDIRECT_TIMEOUT, loopback::wait_for_callback(config.callback_port, config.callback_path))
            .await
            .map_err(|_| Error::OAuthProvider("Timed out waiting for the browser.".into()))?
    }

    /// Exchanges the authorization code for tokens and stores them.
    pub async fn complete(&self, provider: ProviderId, callback: OAuthCallback) -> Result<()> {
        let pkce = self
            .pending
            .lock()
            .unwrap()
            .get(&provider)
            .cloned()
            .ok_or_else(|| Error::OAuthProvider("Start connecting first.".into()))?;
        if callback.state.as_ref().is_some_and(|state| *state != pkce.state) {
            return Err(Error::StateMismatch);
        }
        let config = OAuthConfig::for_provider(provider);
        let (body, _) = self.post_form(config.token_endpoint, &config.exchange_parameters(&callback.code, &pkce.verifier)).await?;
        let bundle = TokenBundle::from_token_response(&body, None, Utc::now(), config.expiry_skew)?;
        self.store(&bundle, provider)?;
        self.pending.lock().unwrap().remove(&provider);
        Ok(())
    }

    pub fn disconnect(&self, provider: ProviderId) -> Result<()> {
        self.cache.lock().unwrap().remove(&provider);
        self.pending.lock().unwrap().remove(&provider);
        keychain::delete_tokens(provider)
    }

    /// A valid token bundle, refreshed if it is about to expire.
    pub async fn active_tokens(&self, provider: ProviderId) -> Result<TokenBundle> {
        let _guard = self.refreshing.lock().await;
        let tokens = self.stored_tokens(provider)?.ok_or(Error::NoProviderConnected)?;
        if !tokens.needs_refresh(Utc::now()) {
            return Ok(tokens);
        }
        let config = OAuthConfig::for_provider(provider);
        let (body, status) = self.post_form(config.token_endpoint, &config.refresh_parameters(&tokens.refresh_token)).await?;
        match TokenBundle::from_token_response(&body, Some(&tokens), Utc::now(), config.expiry_skew) {
            Ok(refreshed) => {
                self.store(&refreshed, provider)?;
                Ok(refreshed)
            }
            // Only a rejected grant means signing in again. An outage must not wipe the saved sign-in.
            Err(error) if status == 400 || status == 401 => {
                let _ = self.disconnect(provider);
                Err(Error::message(format!("{} needs to be reconnected in Settings. {error}", provider.display_name())))
            }
            Err(_) => Err(Error::message(format!(
                "{} sign-in could not be refreshed ({status}). Try again in a moment.",
                provider.display_name()
            ))),
        }
    }

    fn stored_tokens(&self, provider: ProviderId) -> Result<Option<TokenBundle>> {
        if let Some(cached) = self.cache.lock().unwrap().get(&provider) {
            return Ok(Some(cached.clone()));
        }
        let loaded = keychain::load_tokens(provider)?;
        if let Some(tokens) = &loaded {
            self.cache.lock().unwrap().insert(provider, tokens.clone());
        }
        Ok(loaded)
    }

    fn store(&self, bundle: &TokenBundle, provider: ProviderId) -> Result<()> {
        keychain::save_tokens(bundle, provider)?;
        self.cache.lock().unwrap().insert(provider, bundle.clone());
        Ok(())
    }

    async fn post_form(&self, url: &str, parameters: &[(&str, String)]) -> Result<(Vec<u8>, u16)> {
        let response = self
            .http
            .post(url)
            .timeout(REQUEST_TIMEOUT)
            .header("Accept", "application/json")
            .form(parameters)
            .send()
            .await
            .net()?;
        let status = response.status().as_u16();
        Ok((response.bytes().await.net()?.to_vec(), status))
    }
}
